File: packages/schema-domain/src/app-game-adapter-dispatch-preflight.ts
import { AppGameSchemaVersion } from './app-game-primitives';
import {
  AgentAppGameAdapterExecutionDecision,
  AgentAppGameAdapterExecutionState,
  AgentAppGameAdapterHostCapabilityState,
} from './app-game-adapter-execution-readiness';
import { type Infer, NonEmptyStringSchema, Schema, withParser } from './effect';

const DispatchPreflightCount = Schema.Number.pipe(Schema.nonNegative(), Schema.int());

export const AgentAppGameAdapterDispatchPreflightPayloadField = 'appGameAdapterDispatchPreflightReadModel' as const;

export const AgentAppGameAdapterDispatchPreflightState = {
  DispatchEligible: 'dispatch-eligible',
  BlockedBeforeDispatch: 'blocked-before-dispatch',
  ManualRequired: 'manual-required',
  Unavailable: 'unavailable',
  Unsupported: 'unsupported',
  Degraded: 'degraded',
} as const;

export const AgentAppGameAdapterDispatchDecision = {
  DispatchEligible: 'dispatch-eligible',
  BlockedBeforeDispatch: 'blocked-before-dispatch',
} as const;

export const AgentAppGameAdapterDispatchOutcomeState = {
  DispatchReady: 'dispatch-ready',
  ManualRequired: 'manual-required',
  Unavailable: 'unavailable',
  Unsupported: 'unsupported',
  Degraded: 'degraded',
  NotDispatched: 'not-dispatched',
} as const;

const AgentAppGameAdapterDispatchPreflightRowBaseSchema = Schema.Struct({
  schemaVersion: Schema.Literal(AppGameSchemaVersion),
  rowId: NonEmptyStringSchema,
  sourceExecutionReadinessRowId: NonEmptyStringSchema,
  sourceProofEntryId: NonEmptyStringSchema,
  platform: NonEmptyStringSchema,
  productMeanings: Schema.Array(Schema.Literal('native-app', 'native-game')),
  adapterCapability: NonEmptyStringSchema,
  adapterExecutionState: Schema.Literal(
    AgentAppGameAdapterExecutionState.ProvedScopedExecution,
    AgentAppGameAdapterExecutionState.ManualRequired,
    AgentAppGameAdapterExecutionState.Unavailable,
    AgentAppGameAdapterExecutionState.Unsupported,
    AgentAppGameAdapterExecutionState.Degraded,
    AgentAppGameAdapterExecutionState.NotClaimed
  ),
  executionDecision: Schema.Literal(
    AgentAppGameAdapterExecutionDecision.ExecutionAllowed,
    AgentAppGameAdapterExecutionDecision.BlockedBeforeExecution
  ),
  dispatchPreflightState: Schema.Literal(
    AgentAppGameAdapterDispatchPreflightState.DispatchEligible,
    AgentAppGameAdapterDispatchPreflightState.BlockedBeforeDispatch,
    AgentAppGameAdapterDispatchPreflightState.ManualRequired,
    AgentAppGameAdapterDispatchPreflightState.Unavailable,
    AgentAppGameAdapterDispatchPreflightState.Unsupported,
    AgentAppGameAdapterDispatchPreflightState.Degraded
  ),
  dispatchDecision: Schema.Literal(
    AgentAppGameAdapterDispatchDecision.DispatchEligible,
    AgentAppGameAdapterDispatchDecision.BlockedBeforeDispatch
  ),
  dispatchIntentId: Schema.Union(NonEmptyStringSchema, Schema.Null),
  dispatchOutcomeState: Schema.Literal(
    AgentAppGameAdapterDispatchOutcomeState.DispatchReady,
    AgentAppGameAdapterDispatchOutcomeState.ManualRequired,
    AgentAppGameAdapterDispatchOutcomeState.Unavailable,
    AgentAppGameAdapterDispatchOutcomeState.Unsupported,
    AgentAppGameAdapterDispatchOutcomeState.Degraded,
    AgentAppGameAdapterDispatchOutcomeState.NotDispatched
  ),
  dispatchEvidenceRefs: Schema.Array(NonEmptyStringSchema),
  hostCapabilityState: Schema.Literal(
    AgentAppGameAdapterHostCapabilityState.Available,
    AgentAppGameAdapterHostCapabilityState.NotDetected,
    AgentAppGameAdapterHostCapabilityState.NotApplicable
  ),
  hostCapabilityEvidenceRefs: Schema.Array(NonEmptyStringSchema),
  hostCapabilityProbeRefs: Schema.Array(NonEmptyStringSchema),
  dispatchAuditRefs: Schema.Array(NonEmptyStringSchema),
  dispatchTimerRefs: Schema.Array(NonEmptyStringSchema),
  manualProofRequirements: Schema.Array(NonEmptyStringSchema),
  claimBoundary: NonEmptyStringSchema,
  fallbackBehavior: NonEmptyStringSchema,
  adapterDispatchEligible: Schema.Boolean,
  adapterDispatchExecutedClaimed: Schema.Literal(false),
  broadInstalledAppBlockingClaimed: Schema.Literal(false),
  childDeviceDeliveryClaimed: Schema.Literal(false),
  platformEnforcementClaimed: Schema.Literal(false),
  providerDeliveryClaimed: Schema.Literal(false),
  privateDiagnosticsClaimed: Schema.Literal(false),
  lastCheckedAt: NonEmptyStringSchema,
});

type AgentAppGameAdapterDispatchPreflightRowCandidate = Infer<typeof AgentAppGameAdapterDispatchPreflightRowBaseSchema>;

export const AgentAppGameAdapterDispatchPreflightRowSchema = withParser(
  AgentAppGameAdapterDispatchPreflightRowBaseSchema.pipe(
    Schema.filter(
      (row: AgentAppGameAdapterDispatchPreflightRowCandidate) =>
        dispatchPreflightRowIsHonest(row) ||
        'Expected only the scoped Windows owned-process time-limit row to be dispatch-eligible without ' +
          'claiming adapter execution'
    )
  )
);

const AgentAppGameAdapterDispatchPreflightReadModelBaseSchema = Schema.Struct({
  schemaVersion: Schema.Literal(AppGameSchemaVersion),
  readModelId: NonEmptyStringSchema,
  generatedAt: NonEmptyStringSchema,
  sourceReadModelIds: Schema.Array(NonEmptyStringSchema),
  custodyLabel: NonEmptyStringSchema,
  capabilityStatus: NonEmptyStringSchema,
  returned: DispatchPreflightCount,
  dispatchEligibleCount: DispatchPreflightCount,
  blockedBeforeDispatchCount: DispatchPreflightCount,
  adapterDispatchEligibleCount: DispatchPreflightCount,
  adapterDispatchExecutedClaimedCount: Schema.Literal(0),
  hostCapabilityAvailableCount: DispatchPreflightCount,
  hostCapabilityNotDetectedCount: DispatchPreflightCount,
  hostCapabilityNotApplicableCount: DispatchPreflightCount,
  hostCapabilityProbeRefCount: DispatchPreflightCount,
  broadInstalledAppBlockingClaimed: Schema.Literal(false),
  childDeviceDeliveryClaimed: Schema.Literal(false),
  platformEnforcementClaimed: Schema.Literal(false),
  providerDeliveryClaimed: Schema.Literal(false),
  privateDiagnosticsClaimed: Schema.Literal(false),
  rows: Schema.Array(AgentAppGameAdapterDispatchPreflightRowSchema),
});

type AgentAppGameAdapterDispatchPreflightReadModelCandidate = Infer<
  typeof AgentAppGameAdapterDispatchPreflightReadModelBaseSchema
>;

export const AgentAppGameAdapterDispatchPreflightReadModelSchema = withParser(
  AgentAppGameAdapterDispatchPreflightReadModelBaseSchema.pipe(
    Schema.filter(
      (readModel: AgentAppGameAdapterDispatchPreflightReadModelCandidate) =>
        dispatchPreflightReadModelIsConsistent(readModel) ||
        'Expected app/game adapter dispatch preflight counts and row ids to match the rows'
    )
  )
);

export type AgentAppGameAdapterDispatchPreflightRow = Infer<typeof AgentAppGameAdapterDispatchPreflightRowSchema>;
export type AgentAppGameAdapterDispatchPreflightReadModel = Infer<
  typeof AgentAppGameAdapterDispatchPreflightReadModelSchema
>;

function dispatchPreflightRowIsHonest(row: AgentAppGameAdapterDispatchPreflightRowCandidate): boolean {
  return row.dispatchPreflightState === AgentAppGameAdapterDispatchPreflightState.DispatchEligible
    ? dispatchEligiblePreflightRowIsHonest(row)
    : blockedPreflightRowIsHonest(row);
}

function dispatchPreflightReadModelIsConsistent(
  readModel: AgentAppGameAdapterDispatchPreflightReadModelCandidate
): boolean {
  const countExpectations = [
    {
      expected: readModel.dispatchEligibleCount,
      actual: readModel.rows.filter(
        (row) => row.dispatchDecision === AgentAppGameAdapterDispatchDecision.DispatchEligible
      ).length,
    },
    {
      expected: readModel.blockedBeforeDispatchCount,
      actual: readModel.rows.filter(
        (row) => row.dispatchDecision === AgentAppGameAdapterDispatchDecision.BlockedBeforeDispatch
      ).length,
    },
    {
      expected: readModel.adapterDispatchEligibleCount,
      actual: readModel.rows.filter((row) => row.adapterDispatchEligible).length,
    },
    {
      expected: readModel.hostCapabilityAvailableCount,
      actual: readModel.rows.filter(
        (row) => row.hostCapabilityState === AgentAppGameAdapterHostCapabilityState.Available
      ).length,
    },
    {
      expected: readModel.hostCapabilityNotDetectedCount,
      actual: readModel.rows.filter(
        (row) => row.hostCapabilityState === AgentAppGameAdapterHostCapabilityState.NotDetected
      ).length,
    },
    {
      expected: readModel.hostCapabilityNotApplicableCount,
      actual: readModel.rows.filter(
        (row) => row.hostCapabilityState === AgentAppGameAdapterHostCapabilityState.NotApplicable
      ).length,
    },
    {
      expected: readModel.hostCapabilityProbeRefCount,
      actual: readModel.rows.reduce((count, row) => count + row.hostCapabilityProbeRefs.length, 0),
    },
  ] as const;

  return (
    readModel.returned === readModel.rows.length &&
    countExpectations.every(({ expected, actual }) => expected === actual) &&
    new Set(readModel.rows.map((row) => row.rowId)).size === readModel.rows.length
  );
}

function dispatchEligiblePreflightRowIsHonest(row: AgentAppGameAdapterDispatchPreflightRowCandidate): boolean {
  return (
    dispatchEligiblePreflightRowMatchesBoundary(row) &&
    dispatchEligiblePreflightRowTracksDispatch(row) &&
    dispatchEligiblePreflightRowTracksHostCapability(row) &&
    dispatchEligiblePreflightRowKeepsClaimsScoped(row)
  );
}

function blockedPreflightRowIsHonest(row: AgentAppGameAdapterDispatchPreflightRowCandidate): boolean {
  return (
    row.dispatchDecision === AgentAppGameAdapterDispatchDecision.BlockedBeforeDispatch &&
    row.dispatchIntentId === null &&
    row.dispatchOutcomeState !== AgentAppGameAdapterDispatchOutcomeState.DispatchReady &&
    hostCapabilityStateMatchesEvidence(row) &&
    hostCapabilityProbeRefsAreParentSafe(row) &&
    row.manualProofRequirements.length > 0 &&
    !row.adapterDispatchEligible &&
    !row.adapterDispatchExecutedClaimed
  );
}

function dispatchEligiblePreflightRowMatchesBoundary(row: AgentAppGameAdapterDispatchPreflightRowCandidate): boolean {
  return (
    row.platform === 'windows' &&
    row.sourceProofEntryId === 'windows-app-game-owned-process-time-limit' &&
    row.executionDecision === AgentAppGameAdapterExecutionDecision.ExecutionAllowed
  );
}

function dispatchEligiblePreflightRowTracksDispatch(row: AgentAppGameAdapterDispatchPreflightRowCandidate): boolean {
  return (
    row.dispatchDecision === AgentAppGameAdapterDispatchDecision.DispatchEligible &&
    row.dispatchOutcomeState === AgentAppGameAdapterDispatchOutcomeState.DispatchReady &&
    row.dispatchIntentId !== null &&
    row.dispatchEvidenceRefs.length > 0 &&
    row.dispatchAuditRefs.length > 0 &&
    row.dispatchTimerRefs.length > 0
  );
}

function dispatchEligiblePreflightRowTracksHostCapability(
  row: AgentAppGameAdapterDispatchPreflightRowCandidate
): boolean {
  return (
    row.hostCapabilityState === AgentAppGameAdapterHostCapabilityState.Available &&
    row.hostCapabilityEvidenceRefs.length > 0 &&
    row.hostCapabilityProbeRefs.length > 0
  );
}

function dispatchEligiblePreflightRowKeepsClaimsScoped(row: AgentAppGameAdapterDispatchPreflightRowCandidate): boolean {
  return row.manualProofRequirements.length === 0 && row.adapterDispatchEligible && !row.adapterDispatchExecutedClaimed;
}

function hostCapabilityStateMatchesEvidence(row: AgentAppGameAdapterDispatchPreflightRowCandidate): boolean {
  if (row.hostCapabilityState === AgentAppGameAdapterHostCapabilityState.Available) {
    return row.hostCapabilityEvidenceRefs.length > 0;
  }
  if (row.hostCapabilityState === AgentAppGameAdapterHostCapabilityState.NotApplicable) {
    return row.hostCapabilityEvidenceRefs.length === 0;
  }
  return true;
}

function hostCapabilityProbeRefsAreParentSafe(row: AgentAppGameAdapterDispatchPreflightRowCandidate): boolean {
  if (row.hostCapabilityState === AgentAppGameAdapterHostCapabilityState.NotApplicable) {
    return row.hostCapabilityProbeRefs.length === 0;
  }
  return row.hostCapabilityProbeRefs.every((ref) => ref.endsWith('-probe-ref'));
}

File: crates/agent-protocol/tests/contract/app_game_adapter_dispatch_preflight.rs
use ocentra_eventing::expect_value::ExpectValue;
use ocentra_parent_agent_protocol::AppGameAdapterDispatchPreflightReadModel;
use ocentra_parent_agent_protocol::AppGameAdapterDispatchPreflightRow;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_AUDIT_OWNED_PROCESS;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_CLAIM_SCOPED_TIMER;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_DECISION_ELIGIBLE;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_EVIDENCE_OWNED_PROCESS;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_FALLBACK_SCOPED_TIMER;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_INTENT_OWNED_PROCESS_TIME_LIMIT;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_OUTCOME_READY;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_READ_MODEL_ID;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_ROW_ID_PREFIX;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_STATE_ELIGIBLE;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_TIMER_OWNED_PROCESS;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_EXECUTION_DECISION_ALLOWED;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_EXECUTION_ROW_ID_PREFIX;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_EXECUTION_STATE_PROVED_SCOPED;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_HOST_CAPABILITY_AVAILABLE;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_PRODUCT_NATIVE_APP;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_PRODUCT_NATIVE_GAME;
use ocentra_parent_agent_protocol::APP_GAME_PARENT_PLATFORM_WINDOWS;
use ocentra_parent_agent_protocol::APP_GAME_SCHEMA_VERSION;

use ocentra_parent_agent_protocol::constants::v08_supported_adapter_runtime_proof::REF_ADAPTER_CAPABILITY_STATE;

const GENERATED_AT: &str = "2026-06-08T10:16:00Z";
const SOURCE_PROOF_ENTRY_ID: &str = "windows-app-game-owned-process-time-limit";
const ADAPTER_CAPABILITY: &str = "app-game-owned-process-time-limit";

const APP_GAME_ADAPTER_DISPATCH_EXECUTED_CLAIM_FALSE: bool = false;

#[test]
fn app_game_adapter_dispatch_preflight_serializes_parent_safe_rows() {
    let row = AppGameAdapterDispatchPreflightRow {
        schema_version: APP_GAME_SCHEMA_VERSION,
        row_id: format!(
            "{APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_ROW_ID_PREFIX}{SOURCE_PROOF_ENTRY_ID}"
        ),
        source_execution_readiness_row_id: format!(
            "{APP_GAME_ADAPTER_EXECUTION_ROW_ID_PREFIX}{SOURCE_PROOF_ENTRY_ID}"
        ),
        source_proof_entry_id: SOURCE_PROOF_ENTRY_ID.to_string(),
        platform: APP_GAME_PARENT_PLATFORM_WINDOWS.to_string(),
        product_meanings: vec![
            APP_GAME_ADAPTER_PRODUCT_NATIVE_APP.to_string(),
            APP_GAME_ADAPTER_PRODUCT_NATIVE_GAME.to_string(),
        ],
        adapter_capability: ADAPTER_CAPABILITY.to_string(),
        adapter_execution_state: APP_GAME_ADAPTER_EXECUTION_STATE_PROVED_SCOPED.to_string(),
        execution_decision: APP_GAME_ADAPTER_EXECUTION_DECISION_ALLOWED.to_string(),
        dispatch_preflight_state: APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_STATE_ELIGIBLE.to_string(),
        dispatch_decision: APP_GAME_ADAPTER_DISPATCH_DECISION_ELIGIBLE.to_string(),
        dispatch_intent_id: Some(
            APP_GAME_ADAPTER_DISPATCH_INTENT_OWNED_PROCESS_TIME_LIMIT.to_string(),
        ),
        dispatch_outcome_state: APP_GAME_ADAPTER_DISPATCH_OUTCOME_READY.to_string(),
        dispatch_evidence_refs: vec![APP_GAME_ADAPTER_DISPATCH_EVIDENCE_OWNED_PROCESS.to_string()],
        host_capability_state: APP_GAME_ADAPTER_HOST_CAPABILITY_AVAILABLE.to_string(),
        host_capability_evidence_refs: vec![REF_ADAPTER_CAPABILITY_STATE.to_string()],
        host_capability_probe_refs: vec!["windows-host-local-probe-ref".to_string()],
        dispatch_audit_refs: vec![APP_GAME_ADAPTER_DISPATCH_AUDIT_OWNED_PROCESS.to_string()],
        dispatch_timer_refs: vec![APP_GAME_ADAPTER_DISPATCH_TIMER_OWNED_PROCESS.to_string()],
        manual_proof_requirements: vec![],
        claim_boundary: APP_GAME_ADAPTER_DISPATCH_CLAIM_SCOPED_TIMER.to_string(),
        fallback_behavior: APP_GAME_ADAPTER_DISPATCH_FALLBACK_SCOPED_TIMER.to_string(),
        adapter_dispatch_eligible: true,
        adapter_dispatch_executed_claimed: APP_GAME_ADAPTER_DISPATCH_EXECUTED_CLAIM_FALSE,
        broad_installed_app_blocking_claimed: false,
        child_device_delivery_claimed: false,
        platform_enforcement_claimed: false,
        provider_delivery_claimed: false,
        private_diagnostics_claimed: false,
        last_checked_at: GENERATED_AT.to_string(),
    };
    let read_model = AppGameAdapterDispatchPreflightReadModel {
        schema_version: APP_GAME_SCHEMA_VERSION,
        read_model_id: APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_READ_MODEL_ID.to_string(),
        generated_at: GENERATED_AT.to_string(),
        source_read_model_ids: vec![],
        custody_label: "adapter-execution-readiness-and-policy-dispatch".to_string(),
        capability_status: "app-game-adapter-dispatch-preflight-partial".to_string(),
        returned: 1,
        dispatch_eligible_count: 1,
        blocked_before_dispatch_count: 0,
        adapter_dispatch_eligible_count: 1,
        adapter_dispatch_executed_claimed_count: 0,
        host_capability_available_count: 1,
        host_capability_not_detected_count: 0,
        host_capability_not_applicable_count: 0,
        host_capability_probe_ref_count: 1,
        broad_installed_app_blocking_claimed: false,
        child_device_delivery_claimed: false,
        platform_enforcement_claimed: false,
        provider_delivery_claimed: false,
        private_diagnostics_claimed: false,
        rows: vec![row],
    };

    let serialized =
        serde_json::to_value(&read_model).expect_value("dispatch preflight serializes");

    assert_eq!(
        serialized["readModelId"],
        APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_READ_MODEL_ID
    );
    assert_eq!(
        serialized["rows"][0]["dispatchIntentId"],
        APP_GAME_ADAPTER_DISPATCH_INTENT_OWNED_PROCESS_TIME_LIMIT
    );
    assert_eq!(
        serialized["rows"][0]["hostCapabilityEvidenceRefs"][0],
        REF_ADAPTER_CAPABILITY_STATE
    );
    assert_eq!(
        serialized["rows"][0]["hostCapabilityProbeRefs"][0],
        "windows-host-local-probe-ref"
    );
    assert_eq!(serialized["rows"][0]["hostCapabilityState"], "available");
    assert_eq!(serialized["hostCapabilityAvailableCount"], 1);
    assert_eq!(serialized["hostCapabilityProbeRefCount"], 1);
    assert_eq!(
        serialized["rows"][0]["adapterDispatchExecutedClaimed"],
        APP_GAME_ADAPTER_DISPATCH_EXECUTED_CLAIM_FALSE
    );
}
